migration-engine/cli/src/commands.rs

```diff
@@ -68,128 +59,12 @@ enum CliCommand {
-        Some((SqlFamily::Sqlite, path)) => {
-            let path = std::path::Path::new(path);
-            if path.exists() {
-                return Ok(String::new());
-            }
-
-            let dir = path.parent();
-
-            if let Some((dir, false)) = dir.map(|dir| (dir, dir.exists())) {
-                std::fs::create_dir_all(dir)
-                    .context("Creating SQLite database parent directory.")
-                    .map_err(|io_err| CliError::Other(io_err.into()))?;
-            }
-
-            create_conn(datasource, true).await?;
-
-            Ok(String::new())
-        }
-        Some(_) => {
-            let (db_name, conn) = create_conn(datasource, true).await?;
-            conn.create_database(&db_name).await?;
-            Ok(format!("Database '{}' created successfully.", db_name))
-        }
-        None => Err(CliError::Other(anyhow::anyhow!(
-            "Invalid URL or unsupported connector in the datasource ({:?})",
-            url
-        ))),
-    }
-}
-
-/// Try to connect as an admin to a postgres database. We try to pick a default database from which
-/// we can create another database.
-async fn create_postgres_admin_conn(mut url: Url) -> CoreResult<SqlMigrationConnector> {
-    let candidate_default_databases = &["postgres", "template1"];
-
-    let mut params: HashMap<String, String> = url.query_pairs().into_owned().collect();
-    params.remove("schema");
-    let params: Vec<String> = params.into_iter().map(|(k, v)| format!("{}={}", k, v)).collect();
-    let params: String = params.join("&");
-    url.set_query(Some(&params));
-
-    let mut connector = None;
-
-    for database_name in candidate_default_databases {
-        url.set_path(&format!("/{}", database_name));
-        match SqlMigrationConnector::new(url.as_str()).await {
-            // If the database does not exist, try the next one.
-            Err(err) => match &err.kind {
-                migration_connector::ErrorKind::DatabaseDoesNotExist { .. } => (),
-                _other_outcome => {
-                    connector = Some(Err(err));
-                    break;
-                }
-            },
-            // If the outcome is anything else, use this.
-            other_outcome => {
-                connector = Some(other_outcome);
-                break;
-            }
-        }
-    }
-
-    let connector = connector
-        .ok_or_else(|| {
-            ConnectorError::from_kind(ErrorKind::DatabaseCreationFailed {
-                explanation: "Prisma could not connect to a default database (`postgres` or `template1`), it cannot create the specified database.".to_owned()
-            })
-        })??;
-
-    Ok(connector)
-}
-
-fn split_database_string(database_string: &str) -> Option<(&str, &str)> {
-    let mut split = database_string.splitn(2, ':');
-
-    split.next().and_then(|prefix| split.next().map(|rest| (prefix, rest)))
+    let db_name = SqlMigrationConnector::create_database(datasource).await?;
```

Comment:
Yes, I've thought about it. The problem is that with the current connector API, `create_database` takes `&self`, so the constructor cannot rely on the database it is connecting to being present. On the other hand, we want to validate that we have a usable connection in the constructor so we do not forget to validate by mistake.

A consequence of this is that we want a `create_database` method that does not take `&self`, but that would make the whole `MigrationConnector` trait non-object safe (so we couldn't box it anymore, and we do want to box it).

My idea was to introduce a second trait that connectors need to implement, "ManageDatabases" or something like that, for creating databases and possibly other setup-related things.

Reply:
More concretely, the next PR validates that you do not try to migrate the mysql system databases, but we do need to connect to one of these to create another database, so we need `create_database` to skip that validation.

Reply:
Update on this: I double-checked the object safety, and we don't box migration connectors apparently (only MigrationApi). So this could stay in, I'll try that.